### PR TITLE
Single-source system_prompt + first-class prompt overrides (#562)

### LINKED ISSUE #500: Seed surface area: data, training, eval — separate or unified?
## Summary

Surfaced from PR #496 / issue #470 audit. cruijff_kit has multiple stochastic processes in its pipeline, each with its own (or no) seed. There's no single design decision on record about how these relate. The immediate motivator is that **inspect-ai eval has no seed knob at all**, but before plumbing one in we should decide whether eval seeds are a separate concept or part of a unified seed contract.

## Surface area today

| Stage | Seed source | What it controls |
|---|---|---|
| **Data generation** | `data.data_generation.seed` in experiment_summary.yaml (REQUIRED) | Dataset content (sample draws), train/val/test split, perturbation RNG, DGP weight generation (via sub-key `weight_seed` in `model_organisms.rules`). Default 42 in `tabular_to_text_gen` CLI. |
| **Training** | `seed: null` in `src/tools/torchtune/yaml_refs/1B_defaults.yaml`; CLI knob `setup_finetune.py --seed` | torch RNG used by the custom recipes' `training.set_seed()` — weight init (for full-finetune), LoRA init, data shuffling order during epochs, gradient noise where applicable. Default `null` means nondeterministic. |
| **Evaluation** | *(none)* | Would control inspect-ai sampling RNG under `do_sample=True`. |

## Why this needs a design call before implementation

ML reproducibility best practice usually wants seeds **separate per stage** because they ablate different sources of variance:

- `data_seed` fixed, `train_seed` varying: how much does training stochasticity move the metric?
- `train_seed` fixed, `data_seed` varying: how much does sample composition move the metric?
- Both fixed, `eval_seed` varying: how much does eval-time sampling RNG contribute (essentially zero under `temperature=1e-7`, but nonzero in principle)?

A single global seed couples all three, which is simpler to set but **destroys the ability to attribute variance to a stage**. Mattie's recent multi-seed experiments (e.g. `project_multiseed_v_shape_2026-05-09`) used a fixed `data_seed` and varied `train_seed`, which would have been impossible under a single-knob design.

Counter-argument: most users won't think about this and a single seed is one less thing to forget. The advanced case (ablating per-stage) is a small minority.

## Possible directions

1. **Three independent seeds** (`data.data_generation.seed`, training `seed`, `evaluation.seed`). Each propagates to its respective stage; each is independently settable. Matches current data + training behavior; adds an eval seed alongside.

2. **Single `experiment.seed` that fans out** to all three stages by default, but each stage can override. Best of both: simplicity for the common case, control for the advanced case.

3. **Single global seed, no overrides.** Set once, all three stages derive their seeds deterministically from it (e.g. `data_seed = seed`, `train_seed = seed + 1`, `eval_seed = seed + 2`). Loses per-stage ablation.

4. **Status quo + eval seed only.** Just add `evaluation.seed` as a new field; leave data + training as they are. Minimal change, doesn't reconcile the existing inconsistency (REQUIRED data seed vs. default-null train seed).

## Dependencies / coupling

- **Coupled with issue #499 (do_sample workflow)**. If cruijff_kit adopts `do_sample=False` as the eval default, the eval seed becomes moot — greedy has no RNG. The seed work here is most valuable if we keep sampling-based eval, or if we adopt do_sample=False as default but keep sampling as an option.
- **Coupled with experiment reproducibility audits**. Currently no enforced contract that a re-run produces the same results given the same experiment_summary.yaml; this is the path toward enforcing it.

## Acceptance criteria

1. A decision among options 1–4 (or a fourth option not yet on the table).
2. If (1), (2), or (3) is chosen: schema change to experiment_summary.yaml + propagation through `setup_finetune.py` (already partial) and `setup_inspect.py` (would be new). Each `@task` function gets a `seed` parameter and passes it to inspect-ai's `generate()`.
3. If (4) is chosen: just `@task` plumbing + scaffold-inspect agent doc + `TASK_ARG_KEYS`.
4. A test demonstrating that a fixed seed produces a bit-identical eval log on the same hardware.

## Out of scope

- vLLM seed behavior (vLLM has its own seed parameter in SamplingParams; this issue is about the HF provider path).
- Determinism across hardware (CUDA non-determinism is a separate, harder problem).
- Distributed-training seeding semantics (multi-GPU LoRA finetunes have additional considerations beyond a single torch seed).

cc @niznik-dev

### LINKED ISSUE #562: Audit experiment_summary.yaml structure: section principle + duplication
## Summary

Audit the organization of `experiment_summary.yaml`: write down the intended meaning of each top-level section so a field's home is **predictable rather than memorized**, and find/resolve duplication. Surfaced from the #478 discussion about where `dataset_type` belongs.

## Why now

While decoupling eval from torchtune (#478), `dataset_type` turned out to have two plausible homes (`controls` vs `data.training`), and the schema's section boundaries were fuzzy enough that the "right" one wasn't obvious. The immediate case was resolved (see Precedent), but it exposed a general gap: there is no written principle for what belongs in `controls` vs `data` vs `evaluation`, and at least one field is duplicated.

## Current top-level sections

| Section | Apparent intent | Tension |
|---|---|---|
| `experiment` | metadata | — |
| `tools` | tool routing | — |
| `variables` | params swept across runs | — |
| `controls` | params held constant across runs | holds training knobs (epochs, batch_size) **and** task-framing fields also read by eval (prompt, system_prompt, dataset_type) — so it isn't purely "training" |
| `data` | the dataset(s) | — |
| `evaluation` | eval-only knobs | `system_prompt` duplicated here and in `controls` |
| `runs[].parameters` | per-run values | — |

## Known issues to resolve

1. **`system_prompt` duplication** — lives in both `controls.system_prompt` and `evaluation.system_prompt`, kept equal by convention ("must match training") rather than by construction. A single source with explicit per-cell overrides would remove the drift risk.
2. **No written principle** for section membership — contributors memorize where fields go instead of predicting it.

## Precedent from #478

`dataset_type` belongs in `controls`, not `data.training`, because it is an **experiment-wide invariant** (training / validation / eval must share the same format) rather than a per-split data property — nesting it under `data.training` would falsely imply it could differ by split. This "is it a constant, or a per-X property?" test is exactly the kind of reasoning this audit should standardize.

## Relationship to existing issues

- **#500** (seed surface area) — the same "one home or per-stage?" question, scoped to seeds. This issue generalizes it to the whole schema; resolve them consistently.
- **#372** (naming audit) — adjacent but orthogonal: #372 is about what fields are *named*; this is about *where they live* and *duplication*.

## Acceptance criteria (audit-first)

1. A written, documented principle for each top-level section's membership (template and/or ARCHITECTURE docs).
2. An inventory of every field, its home, and whether that home is semantically correct.
3. A decision on `system_prompt` de-duplication (single source + overrides vs status quo).
4. Cross-check with #500 and #372 so the three land a consistent schema story.

Audit-first: implementation/migration can follow once @sarahepedersen / @msalganik weigh in on the principle, same as #372.

### PR DESCRIPTION
**Pull request by niznik-dev**
Closes #562

 🎻 **MxC** ✳️

# Description

The #562 audit was *audit-first* pending @sarahepedersen / @msalganik sign-off. With them away and the proposed changes small, this PR implements the audit so the shape is reviewable now — it's the capstone before the release + a development pause. The principle in `ARCHITECTURE.md` is still open to their comment when they're back; nothing here forecloses that.

The audit found **one** real code change (A) and **one** doc addition (B). Section **C** is a deliberate scope extension — wiring prompt variability — added because it's the last thing before the pause and the system_prompt collapse made it cleaner, not harder. C is **zero production-logic change**: it rides the same override-wins propagation A leans on.

## A — Collapse `system_prompt` to a single source (breaking)

`experiment_summary.yaml` carried the system prompt in two homes — `controls.system_prompt` (→ training) and `evaluation.system_prompt` (→ eval) — kept equal only by the template instruction *"must match training."* Convention, not construction → silent-drift risk.

- `EVAL_FIELDS` now propagates **`controls.system_prompt`** to eval, alongside the other task-framing invariants it already carries (`prompt`, `dataset_type`). This makes the propagation map itself more principled: task-framing comes from `controls`, eval-only knobs from `evaluation`.
- `evaluation.system_prompt` is **removed** from the template.
- Per-task variation stays at **`evaluation.tasks[].system_prompt`** (a legitimate per-cell override, e.g. cue vs. no-cue ablations).
- New regression test locks the collapse: a stray `evaluation.system_prompt` is **not** a source and cannot override the single home.

**Migration:** delete `evaluation.system_prompt` from existing files — its value should already equal `controls.system_prompt`, now the only home. A stray one is silently ignored.

## B — Document the section-membership principle (additive)

New `ARCHITECTURE.md` section: a field's home is decided by **granularity × role**, with two boundary rules and a duplication-vs-independence test.

- **`controls` is not "training-only"** — it holds experiment-wide *invariants*, whichever stage consumes them (generalizes the #478 `dataset_type` ruling).
- **A field both train and eval need lives in `controls`, never duplicated into `evaluation`.**
- **Duplication vs. independence:** must-match values collapse (`system_prompt`); legitimately-independent per-stage values stay (`seed` — resolves #500 in place, no change).
- Names the **three** distinct "system prompt" surfaces (generation `context`, training/eval `controls.system_prompt`, per-cell override).

## C — Make `prompt` a first-class per-task / per-run override (scope extension)

The user `prompt` could only be set experiment-wide at `controls.prompt`. But the real prompt-engineering workflow — sweep one appended sentence across N conditions — had **no supported home**. Past experiments (#539 spike, base-rate ladder) hand-patched each cell's `eval.yaml` *after* scaffolding: an undocumented, unvalidated step between scaffold and run. Meanwhile the schema offered a per-task override for `system_prompt` (held *constant* in practice) but not for `prompt` (the actual workhorse) — the flexibility was inverted relative to need.

Two override surfaces, both riding the existing idempotent override-wins propagation:

- **`evaluation.tasks[].prompt`** — eval-side prompt sweep (one task script, N tasks differing only by an appended sentence). The zero-finetuning prompt-engineering pattern, now first-class instead of hand-patched.
- **`runs[].parameters.prompt`** — a fine-tune trains *and* evaluates on its own prompt. Flows to training via `setup_finetune.yaml` (the `config_file > argparse-default` merge already carried it — verified by a new test) and is carried into the run's eval cells for **train/eval parity**.

Precedence at a cell: **per-task > per-run > `controls.prompt`**.

Why `prompt` *layers* where `system_prompt` *collapsed*: they aren't peers. `prompt` is the per-sample user-message template (task content — instructions, hints, appended sentences), applied to chat *and* base models, and empirically the high-frequency lever. `system_prompt` is a chat-only framing held constant. `ARCHITECTURE.md` now records this distinction beside the surfaces note.

And `design-experiment` now **teaches** it (not just exposes the field): a **Prompt Sweeps** recipe in `param_selection.md` recognizes framing / appended-sentence / cue conditions and authors the structure — Pattern A (eval-only: one `tasks[]` entry per condition, same script, short name + per-task `prompt`, fanned across the matrix) or Pattern B (per-run `parameters.prompt`). So the end-to-end path is guided, not hand-edited.

**No migration:** purely additive. `controls.prompt` behaves exactly as before; the overrides are opt-in.

## Deferred (not in this PR)
- `output` (single-field `wandb_project`) → fold into `experiment`? Parked, low priority.
- `torchtune_config.yaml` vs `finetune.yaml` dual-naming — a naming question (#372 axis), route separately.
- Issue's section table omits `models` / `output` — noted; the principle covers all 9.

## New Dependencies
None.

# Testing Instructions
- Full unit suite green: `pytest tests/unit` (1245 passed). New coverage:
  - `TestPromptFlow` (`test_setup_finetune_main.py`) — `controls`/per-run prompt reaches the training dataset; CLI > config_file > default precedence.
  - `test_eval_per_task_prompt_override_beats_controls_source` + `test_train_per_run_prompt_override_survives` (`test_propagate.py`) — agent-written prompt survives propagation on both sides.
  - plus the system_prompt collapse + warn-on-stray tests from A.
- `ruff check . && ruff format --check .` clean.

—MxC

### CODE REVIEW

File: .claude/skills/design-experiment/templates/experiment_summary.yaml [hunk 1]
```diff
@@ -203,7 +212,9 @@ runs:
 # =============================================================================
 
 evaluation:
-  system_prompt: string            # REQUIRED: Must match training system prompt
+  # NOTE: the system prompt is NOT set here — it has a single source at
```
Comment (niznik-dev) [inline .claude/skills/design-experiment/templates/experiment_summary.yaml]: This seems unnecessary

File: src/tools/experiment/propagate.py [hunk 2]
```diff
@@ -23,18 +24,20 @@
...
     "evaluation.do_sample": "do_sample",
     "evaluation.max_tokens": "max_tokens",
     "evaluation.max_connections": "max_connections",
     "evaluation.scorers": "scorers",
-    # The @task reads `prompt` from config_path at runtime (default "{input}")
-    # and the agent derives `use_chat_template` from `dataset_type`. Both lived
-    # only in setup_finetune.yaml before; carrying them here lets an eval-only
-    # run (no fine-tuning, no setup_finetune.yaml) source them from the
-    # experiment_summary single source of truth instead of silently defaulting.
+    # Task-framing invariants live in `controls` (one home, shared by training
```
Comment (niznik-dev) [inline src/tools/experiment/propagate.py]: A bit wordy

File: src/tools/experiment/propagate.py [hunk 3]
```diff
@@ -99,12 +102,26 @@ def propagate_eval_fields(experiment_summary: dict, eval_config: dict) -> dict:
...
     Idempotent: existing non-None values in eval_config take precedence so the
-    agent's per-cell decisions (e.g. per-task system_prompt overrides) survive
-    propagation.
+    agent's per-cell decisions (e.g. per-task prompt / system_prompt overrides,
+    or a per-run prompt carried in for train/eval parity) survive propagation.
 
     The eval seed is resolved (evaluation.seed, else DEFAULT_SEED) and written
     unless the agent already set a per-cell seed, which wins.
+
+    Warns on a stray top-level ``evaluation.system_prompt``: it is no longer a
```
Comment (niznik-dev) [inline src/tools/experiment/propagate.py]: Shorten since the code is self-explanatory

File: CHANGELOG.md [hunk 4]
```diff
@@ -7,6 +7,7 @@ All notable changes to cruijff_kit will be documented in this file.
...
+- **Per-task and per-run prompt overrides**: the user `prompt` can now vary as a first-class field, not just the experiment-wide `controls.prompt`. Set `evaluation.tasks[].prompt` for an eval-side prompt sweep (N tasks, one script, a different appended sentence each — the zero-finetuning prompt-engineering pattern, previously done by hand-patching each cell's `eval.yaml`), or `runs[].parameters.prompt` to give a fine-tune its own prompt across both training and its eval cells (train/eval parity). Both ride the existing override-wins propagation; precedence is per-task > per-run > `controls.prompt`. Any prompt (whatever its source) is now checked for the `{input}` placeholder at scaffold/eval time and fails loudly — `str.format` would otherwise drop the record silently and run every sample on the same input-less text. (#562)
```
Comment (niznik-dev) [inline CHANGELOG.md]: Shorten

File: CHANGELOG.md [hunk 5]
```diff
@@ -36,6 +37,7 @@ Each requires migrating existing config files; see the migration note per item.
...
 - **Deterministic `experiment_summary.yaml` propagation** — `propagate_*_fields()` is a mandatory propagate-first step in scaffolding, so generated `eval_config.yaml` / `setup_finetune.yaml` derive from one verified source. (#502)
 - Evaluation scaffolding (`scaffold-inspect`) no longer reads `setup_finetune.yaml` — `prompt` and `dataset_type` are propagated from `experiment_summary.yaml`, decoupling eval from the training artifact. (#478)
+- **`system_prompt` is now single-sourced at `controls.system_prompt`** — the duplicate `evaluation.system_prompt` (kept equal to training only by convention) is removed; eval derives it via propagation, alongside the other task-framing invariants `prompt` and `dataset_type`. Per-task eval variation stays at `evaluation.tasks[].system_prompt`. **Breaking:** delete `evaluation.system_prompt` from existing files; its value should already equal `controls.system_prompt`, which is now the only home (and is now `REQUIRED`). A leftover `evaluation.system_prompt` is no longer read and `propagate_eval_fields` **warns** when it sees one, so an un-migrated file can't silently mismatch. (#562)
```
Comment (niznik-dev) [inline CHANGELOG.md]: Shorten

File: CHANGELOG.md [hunk 6]
```diff
@@ -48,6 +50,7 @@ Each requires migrating existing config files; see the migration note per item.
...
 - **Tool-domain folder convention + `inspect/`-shadows-stdlib hazard** documented (`docs/ARCHITECTURE.md`, `src/tools/inspect/__init__.py`) and locked by a guard test (`tests/unit/test_inspect_no_stdlib_shadow.py`). Stale `ck-experiments/` paths fixed in the `create-quiz` docs and a `summarize` test fixture. (#372)
+- **`experiment_summary.yaml` section-membership principle written down** (`docs/ARCHITECTURE.md`): a field's home is decided by granularity × role, with two boundary rules (`controls` holds experiment-wide invariants regardless of consuming stage; a field both train and eval need lives in `controls`, never duplicated into `evaluation`) and a duplication-vs-independence test (`system_prompt` collapses; `seed` stays per-stage). Names the three distinct "system prompt" surfaces. (#562)
```
Comment (niznik-dev) [inline CHANGELOG.md]: Shorten
